Integrate various fixes and updates to main

acraugh opened the pull request:
<!--
  Title ^^^^ above ^^^^ should follow good commit message best practices wherever possible.

  A properly formed git commit subject line should always be able to complete the following sentence:

    If applied, this commit will <your subject line here>
-->
Various fixes and checks to bring multiple revision branches into a common set.

## Summary
<!-- Brief summary of changes if not sufficiently described by commit messages. -->
Typo corrections, documentation updates, regression test updates, and a review of new elements added to the IngestLDD. Note that additional action is needed for the IngestLDD definitions marked "INADEQUATE" before the branch can be merged and the dictionary released.

## Test Data and/or Report
<!--
  One of the following should be included here:
    * Reference to regression test included in code (preferred wherever reasonable)
    * Attach test data here + outputs of tests
-->
See test/

## Related Issues
<!--
  Reference related issues here and use `Fixes` or `Resolves` for closing issues, for example:
  * Refs pds-data-dictionaries/PDS4-LDD-Issue-Repo#13 will include a link between this PR and that issue
  * Resolves pds-data-dictionaries/PDS4-LDD-Issue-Repo#13 will actually close the ticket once this PR is merged
-->
Dictionary is in active development - no issues filed.

cgobat committed:
Add detectors/types, phase names, & engineering conversions; generalize conversion constant class

pdsen-ci committed:
Add logs from failed LDD generation

cgobat committed:
Increment minor instead of major version

pdsen-ci committed:
Add logs from failed LDD generation

cgobat committed:
Update unit test labels

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Permit KEM1 & KEM2 cruise phases; add optional `arrokoth_constant`

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Fix document LID references; de-duplicate class names

cgobat committed:
Update namespace outline

cgobat committed:
MVIC_Conversion_Constants -> Radiometric_Conversion_Constants

cgobat committed:
Correct document bundle and collection names in reference list

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Fix typo: electrostaic -> electrostatic

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Make `sweep_samples_count` optional as it only applies to some SWAP products

cgobat committed:
Add SWAP mockup label doc page

cgobat committed:
Add valid SWAP test label for CI pipeline

pdsen-ci committed:
Add logs from failed LDD generation

cgobat committed:
Remove reference to currently nonexistent context product LID

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Include SWAP mockup page in ToC

cgobat committed:
Add spacecraft state information

cgobat committed:
Add LEISA and Ralph (common) details

cgobat committed:
Remove trailing whitespace

pdsen-ci committed:
Add logs from failed LDD generation

cgobat committed:
Add spacecraft state class to test labels

pdsen-ci committed:
Add logs from failed LDD generation

cgobat committed:
Switch from enumerated "enabled"/"disabled" to booleans

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Reorder details classes

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Change `true510` attribute to a boolean

cgobat committed:
Report LEISA temp in deg C rather than DN

cgobat committed:
Allow complete set of possible pointing methods and spin states

cgobat committed:
Add a LEISA mockup label

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Add a missed period in filename

cgobat committed:
Add LEISA mockup to ToC

cgobat committed:
Update outline with added classes to date

acraugh committed:
Update PDS4_NH_IngestLDD.xml

Fixed "Expriment" type

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

acraugh committed:
Merge pull request #11 from pds-data-dictionaries/proposed-changes

First round of changes proposed by mission team

Merged into "fixup" branch for additional cleanup in docs/, test/, and IngestLDD

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

acraugh committed:
Merge pull request #12 from pds-data-dictionaries/feature/spacecraft_and_leisa_info

Add spacecraft state information and Ralph/LEISA details

Merged into "fixup" branch for cleanup of docs/, test/, and IngestLDD

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

acraugh committed:
Various fixes and editorial updates

Updates to schema references in mockup.
Check and marking of IngestLDD descriptions that are not sufficient; fixed a number of definitions; typo correction; uniform style of definition; uniform use of "nh:" prefix to referencce dictionary elements.
Added deprecation regression tests.

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -6,15 +6,52 @@
+    Initials/handles:
+       ACR        Anne Raugh, SBN at UMD
+       C. Gobat   Caden Gobat, SWRI
+       
```
Should add `MKC        M. Kate Crombie, Indigo Information Systems, LLC` for completeness

CC @katecrombie

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -49,20 +86,85 @@
+        </value_meaning>
+      </DD_Permissible_Value>
+      <DD_Permissible_Value>
+        <value>Long-Range Reconnaissance Imager</value>
```
Note: this conflicts with 41e1af85df964a9accebeac2a9009f483f4ef575, which is part of what we were trying to avoid by doing this development collaboratively in close-to-real time... But everyone needs to be on board with that strategy in order for it to work.

We will need to sort such things out.

acraugh replied:
I don't understand why this is a conflict. This appears to reference something in another development branch, not something included in this PR. This is what the biweekly (perhaps they should be weekly?) meetings are supposed to sort out, along with us reviewing each others PRs. Since that branch only appeared yesterday, there hasn't been a meeting yet in which to discuss it...

cgobat replied:
It's a conflict in the sense that this PR contains LORRI content that is different than what was added in the other new branch. If/when we try to merge both things into the same place, there will be a merge conflict and we will have to pick one or the other.

acraugh replied:
Noted, and Ben has gotten his heads-up. The value will remain the same, but aspects of the definitions might be merged. That will be a separate action after this merge, though.

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -114,37 +216,153 @@
+        </value_meaning>
+      </DD_Permissible_Value>
+      <DD_Permissible_Value>
+        <value>Ultraviolet Imaging Spectrograph</value>
```
Same comment as above for LORRI regarding 41e1af8

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -114,37 +216,153 @@
       <DD_Permissible_Value>
         <value>CCD</value>
-        <value_meaning>The detector is a standard CCD.</value_meaning>
+        <value_meaning>INADEQUATE [How is this different from the "Frame Transfer CCD" following?] The detector is a standard CCD.</value_meaning>
```
"CCD" was already present for MVIC. It seems we added "Frame Transer CCD" for LORRI, maybe having glossed over the fact that there was already a CCD value available.

@katecrombie: are you on board with striking "Frame Transfer CCD" and using "CCD" for everything with a CCD?

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -114,37 +216,153 @@
+      </DD_Permissible_Value>
+      <DD_Permissible_Value>
+        <value>Dust Imapct Counter</value>
+        <value_meaning>INADEQUATE [spell out "PVDF"] The detector is a set of permanently polarized PVDF plastic films      
```
Will do

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -114,37 +216,153 @@
+      <DD_Permissible_Value>
+        <value>HgCdTe PICNIC array</value>
+        <value_meaning>
+          INADEQUATE [spell out PICNIC] The detector is a modified HgCdTe PICNIC array.</value_meaning>
```
PICNIC isn't an acronym (or at least if it is, I've never seen it spelled out). It's the name of a specific detector/product made by Rockwell. Some info [here](https://people.ast.cam.ac.uk/~optics/tech/hawaii/picnic.htm) and [here](https://web.archive.org/web/19970222122800/http://creation.risc.rockwell.com:80/779/mct_fpa/picnic.html)

acraugh replied:
So it is. It is also specific to Rockwell (that is, a trademark). Any chance we could tie it to "Rockwell Scientific Imaging"? Mainly this is for users generations from now who might want a clue for researching the instrument.

cgobat replied:
Okay, so just change it to "Rockwell Scientific Imaging PICNIC HgCdTe array"?

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -211,203 +445,237 @@
-      The nh:pluto_constant attribute supplies the constant to be used for sources with 
-      spectral distribution similar to (134340) Pluto to convert MVIC calibrated data to
-      physical units.
+      INADEQUATE [Units are wrong for a rate, which would be (time^-1, not time). Exposure time is usually specified as am img: class unless there are special considitions that would be described here.] 
```
This is in fact a duration, not technically a rate, so units of time are correct. It is slightly different than exposure time (in principle at least, if not in practice)&mdash;it is defined as the time between readouts, not the time the sensor was exposed for.

I will update this description.

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -416,7 +684,210 @@
+    <nillable_flag>false</nillable_flag>
+    <submitter_name>MKC</submitter_name>
+    <definition>
+      INADEQUATE [the pattern implies encoded information. How should the value be interpreted?] 
```
Will add details to definition

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -211,203 +445,237 @@
-      The right_side_median_bias attribute provides the median bias, in DN,
-      determined by averaging the bias levels for the shielded pixels on 
-      the right side of the detector array.
+      [INADEQUATE Why is the minimum negative?] The nh:met510 attribute provides the mission elapsed time (MET) of the Ralph housekeeping packet that marks the start of an observation, used to determine the observation start time and frame rate.
```
@katecrombie do you remember why this had a minimum of -1?

cgobat replied:
This might just be a copy/paste issue originating in `nh:(left|right)_side_median_bias`.

acraugh replied:
We can leave a flag here, if you like ("[REPLACE]" or maybe "[CHECK]"?) and move on...

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -416,7 +684,210 @@
+    <nillable_flag>false</nillable_flag>
+    <submitter_name>MKC</submitter_name>
+    <definition>
+      INADEQUATE [The attribute name should be more specific so that the boolean values can be immediately interpreted. For example, "relative_control_mode_active" would be better.]
```
Makes sense. Will change.

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -468,20 +965,52 @@
+    <nillable_flag>false</nillable_flag>
+    <submitter_name>MKC</submitter_name>
+    <definition>
+      INADEQUATE [Replace "TBR" with something meaningful.] The nh:target_motion_rate attribute is the rate of motion of the target with respect to the instrument detector in units of (TBR). 
```
Will fix

cgobat replied:
I am currently waiting to hear back from the Ralph instrument team about what units these `*RATE` keywords are in.

acraugh replied:
If we're likely to be waiting on this for awhile I'm inclined to replace the flag with [INCOMPLETE] and move on. This is the last thing to be resolved prior to merge, as far as I can see.

acraugh replied:
In fact, I just did that and also rewrote the "coefficient_*" definitions to include context and removed those flags. That clears everything and we're good to merge as far as I'm concerned. I am going to hold off for an hour or so to compose a document for the src/ directory on conventions and style for the IngestLDD so we can have something reasonably consistent. If it works well we can offer it to the rest of PDS.

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -538,27 +1067,88 @@
+    <nillable_flag>false</nillable_flag>
+    <submitter_name>MKC</submitter_name>
+    <definition>
+      INADEQUATE [The attribute name should make the boolean value meaningful, e.g., "nh:thruster_x_enabled"] The nh:thruster_y_enable attribute describes the state of the z thruster. The state is either "enabled" or "disabled"
```
Okay. I can add a "d" to the ends of these attribute names.

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -538,27 +1067,88 @@
+    <nillable_flag>false</nillable_flag>
+    <submitter_name>MKC</submitter_name>
+    <definition>
+      INADEQUATE [What is a "0x150 packet"? The attribute name should make the meaning of the value obvious, e.g., nh:housekeeping_packet_is_real] The nh:true510 attribute describes whether the 0x510 Ralph housekeeping packet is real (true) or assumed from a gap (false)
```
Will revise

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -173,6 +391,22 @@
+    <nillable_flag>false</nillable_flag>
+    <submitter_name>MKC</submitter_name>
+    <definition>
+      INADEQUATE [what is "XXXXX"?] The nh:gc_scan_rate attribute provides the scan rate of XXXXX in units of TBD
```
Will fix

cgobat commented on src/PDS4_NH_IngestLDD.xml:
```diff
@@ -432,14 +903,40 @@
+      <DD_Permissible_Value>
+        <value>3-Axis</value>
+        <value_meaning>
+          INADEQUATE [No "TBDs" allowed. Ever.] TBD
```
I will try to get definitions for these now, but the fact is that TBDs are sometimes unavoidable as a temporary measure. As a mission, we cannot afford (schedule-wise) to put the development of processing and archive preparation software (which depends on established structures and content of things like LDD attributes and allowable values) on hold until publication-ready definitions are produced. While it would be ideal to have fully fleshed-out documentation in place ahead of time, this is not always (maybe never?) the case and delivery deadlines would never be met if software development had to wait for all such things to be finalized.

Yes, TBDs need to be addressed prior to release. It's fairly simple to Ctrl-F "TBD"/"TBS"/"TBR" once things have settled to make sure all of our ducks are sufficiently in a row for public consumption.

It's completely fine with Kate and I if you want to hold things in a staging/fixup/development branch until ready to go, but we need to keep pushing forward on our side. I will look to see if we have anything on hand to inform a definition for these, but no promises.

cgobat replied:
I could make something up for now, e.g. "Spinning" &rarr; "Spacecraft is spinning" and "3-Axis" &rarr; "Spacecraft pointing is triaxially stable", but the way I see it, that is potentially worse than a TBD because the intent to revisit it later is less clear and it would be much easier for something like that that we *did* want to change to end up slipping through the cracks.

acraugh replied:
Another place where a flag (probably "[REPLACE]" for this) can be left.

katecrombie replied:
How about we use [INCOMPLETE] as our flag for definitions that we know we need, but have incomplete or inadequate items?  I like [CHECK] as the flag if we are unsure of the facts given.  That way we KNOW we need to search these two terms and ajudacate any that are remaining prior to LDD publication.

cgobat committed:
Update submitter identifiers

cgobat committed:
Remove duplicate "Frame Transfer CCD" permissible value

cgobat committed:
Spell out PVDF

cgobat committed:
Expand definition of `gc_scan_rate` attribute

cgobat committed:
Clarify `leisa_rate` attribute definition

cgobat committed:
Expand definition of `pointing_method` to explain what the various possible values mean

cgobat committed:
Update attribute names to make the meaning of boolean values more apparent

cgobat committed:
Update attribute names in docs and tests to match

cgobat committed:
Specify the manufacturer of PICNIC arrays

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat committed:
Fix typo "SAWP" -> "SWAP"

acraugh commented:
So the problem is, as you demonstrated in an earlier comment, that while I
might think to search for "TBD", how the heck would I know to also search
for "TBR" and whatever the third one was that I can't even remember long
enough to complain about it?

The definitions don't have to be "publication-ready". They have to be good
enough that when the updates are forgotten the reviewers and users do not
find a dictionary full of "TBD" abbreviations. Partial information is
better than none.

For marking things needing update, I would suggest we settle on a string
that will always be used regardless of context - like, say "[FINISH]" or
"[REPLACE]". That is something that is typed intentionally and we can then
know for certain what ONE string to search on to find those things needing
additional detail. No matter how good the intentions, I find things like
"TBD" in published documentation on a regular basis in the PDS. Schedules
are too tight - the pieces never get picked up. A visually distinctive flag
("TBD" is very effectively camouflaged in a document already stuffed to the
gills with acronyms) might at least be a constant visual annoyance that
serves as a reminder.

So, for example, for something like the spin state values, you might have a
placeholder definition like: "[REPLACE] Spin states and their identifiers
will be defined in the Instrument SIS,...." and then name a section or give
a document number. Then, at least, when it is forgotten a seriously
interested user has half a chance of finding the information they need
somewhere. And yes, it might be the same for each value in a case like this.

As for the conflicting changes, I don't recognize them because other folks
introduced them. What's the most efficient way to solve this? We have a
meeting scheduled next week, but that's a bit far off. As far as deadlines
go, ours is in May - so this needs to be solved relatively quickly. We'd
need to change and rerun templates and validation, plus edit any associated
documentation.

I'm out for an urgent doctor's appointment this morning from about 10am-1pm
EDT. I'll try to pull folks in for a 2pm EDT meeting today - it's our usual
time, just a week early. Let's see who's available.

Thanks!

-Anne.


On Wed, Apr 3, 2024 at 8:49 PM Caden Gobat ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/PDS4_NH_IngestLDD.xml
> <https://github.com/pds-data-dictionaries/ldd-nh/pull/14#discussion_r1550684301>
> :
>
> > @@ -432,14 +903,40 @@
>      </DD_Value_Domain>
>    </DD_Attribute>
>
> +  <DD_Attribute>
> +    <name>spacecraft_spin_state</name>
> +    <version_id>1.0</version_id>
> +    <local_identifier>spacecraft_spin_state</local_identifier>
> +    <nillable_flag>false</nillable_flag>
> +    <submitter_name>MKC</submitter_name>
> +    <definition>
> +      The nh:spacecraft_spin_state attribute indicates which of the spacecraft spin states was employed during the observation sequence.</definition>
> +    <DD_Value_Domain>
> +      <enumeration_flag>true</enumeration_flag>
> +      <value_data_type>ASCII_Short_String_Collapsed</value_data_type>
> +      <DD_Permissible_Value>
> +        <value>3-Axis</value>
> +        <value_meaning>
> +          INADEQUATE [No "TBDs" allowed. Ever.] TBD
>
> I could make something up for now, e.g. "Spinning" → "Spacecraft is
> spinning" and "3-Axis" → "Spacecraft pointing is triaxially stable", but
> the way I see it, that is potentially worse than a TBD because the intent
> to revisit it later is less clear and it would be much easier for something
> like that that we *did* want to change to end up slipping through the
> cracks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pds-data-dictionaries/ldd-nh/pull/14#discussion_r1550684301>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADBV2QFPTSMHUZFTZMNCO7TY3SPS7AVCNFSM6AAAAABFWA77S2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNZYGM2TQOJUGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

acraugh commented:
Yes, but that's a conflict that will exist then, not now.  Is it a reason
to hold up this merge?  More specifically, are you planning to make use of
those particular additions immediately?

-Anne.


On Thu, Apr 4, 2024 at 11:29 AM Caden Gobat ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/PDS4_NH_IngestLDD.xml
> <https://github.com/pds-data-dictionaries/ldd-nh/pull/14#discussion_r1551919012>
> :
>
> >      </definition>
>      <DD_Value_Domain>
>        <enumeration_flag>true</enumeration_flag>
>        <value_data_type>ASCII_Short_String_Collapsed</value_data_type>
> +      <DD_Permissible_Value>
> +        <value>Linear Etalon Imaging Spectral Array</value>
> +        <value_meaning>
> +          Data were obtained using the RALPH/LEISA infrared detector with a waveleght range of 1250 - 2500 nm.
> +        </value_meaning>
> +      </DD_Permissible_Value>
> +      <DD_Permissible_Value>
> +        <value>Long-Range Reconnaissance Imager</value>
>
> It's a conflict in the sense that this PR contains LORRI content that is
> different than what was added here. If we try to merge both things into the
> same place, there will be a merge conflict and we will have to pick one or
> the other.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pds-data-dictionaries/ldd-nh/pull/14#discussion_r1551919012>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADBV2QFWULLUYX6B6S342RTY3VWTZAVCNFSM6AAAAABFWA77S2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSOBQGI4TGMBRHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

cgobat commented:
No, it's not a reason to hold this up. I was just flagging it as something that will need to be discussed/resolved sooner or later.

cgobat committed:
Rename `true510` attribute -> `hk_packet_is_real`

cgobat committed:
Add [CHECK] on `met510` regarding negative minimum

cgobat committed:
Update PICNIC array description

cgobat committed:
Add preliminary S/C spin state descriptions with flags to revisit

cgobat committed:
Rename `true510` -> `hk_packet_is_real` in docs too

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

acraugh committed:
Update PDS4_NH_IngestLDD.xml

Modified nh:coefficient_* definitions and replace and replaced nh:target_motion_rate flag with [INCOMPLETE].

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

acraugh committed:
Add Developer's notes

Update IngestLDD change history; add developers notes to src/.

acraugh committed:
Merge branch 'fixup' of https://github.com/pds-data-dictionaries/ldd-nh into fixup

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

acraugh committed:
Update Notes_for_Developers.md

More notes...

acraugh committed:
Merge branch 'fixup' of https://github.com/pds-data-dictionaries/ldd-nh into fixup

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

acraugh committed:
Update ChangeLog.md

Added Notes_for_Developers doc (#17)

acraugh committed:
Update Notes_for_Developers.md

Fixing typos

acraugh committed:
Merge branch 'fixup' of https://github.com/pds-data-dictionaries/ldd-nh into fixup

pdsen-ci committed:
Auto-generated LDDs by PDSEN CI Bot

cgobat reviewed (approved):
Good to go
